### DISCUSSION TITLE
Visibility change callbacks for `<Disclosure>`?

### DISCUSSION THREAD

**Discussion by SevenOutman** [tailwindlabs/headlessui#678] [Help]
Hi, I wonder if there's any callback props on `<Disclosure>` that fires when it expands/collapses?

```jsx
<Disclosure onShow={} onHide={}>
  ...
</Disclosure>

**Comment by 0x7357** [comment]
+1

That's the exact same question I wanted to ask. I'd like to use Disclosure with a callback to load the DisclosurePanel content via (e. g.) Axios.

**Reply by SevenOutman** [reply]
> I'd like to use Disclosure with a callback to load the DisclosurePanel content via (e. g.) Axios.

I think this use case can be solved with a `useEffect` depending on `open` render prop:

```jsx
function MyAsyncComponent({ shown }) {
  useEffect(() => {
    if (shown) {
      // load the data
    }
  }, [shown]);
}

<Disclosure>
  // ... Button
  <Disclosure.Panel>
    {({ open }) => (
      <MyAsyncComponent shown={open} />
    )}
  </Disclosure.Panel>
</Disclosure>
```

**Reply by 0x7357** [reply]
I guess this would work, but isn't as clean as firing a callback, like onShow/onHide. But thanks for the idea!
